File: leetcode/94_binary_tree_inorder_traversal.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def inorderTraversal(self, root):
        result = []
        if not root:
            return result
        stack = []
        while stack or root:
            while root:
                stack.append(root)
                root = root.left
            curr = stack.pop()
            result.append(curr.val)
            root = curr.right
        return result

File: leetcode/java/94BinaryTreeInorderTraversal.java
/**
 * T: O(N) S: O(N)
 *
 * Traverse the left subtree until the end on the each iteration. Start with the deepest left-most node and walk back
 * from it, visiting its parent and right child. */
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class TreeNode {
  int val;
  TreeNode left;
  TreeNode right;

  TreeNode(int x) {
    val = x;
  }
}

class Solution {
  public List<Integer> inorderTraversal(TreeNode root) {
    List<Integer> traversedNodes = new ArrayList<>();
    Stack<TreeNode> nodesToProcess = new Stack<>();
    while (!nodesToProcess.isEmpty() || root != null) {
      while (root != null) {
        nodesToProcess.push(root);
        root = root.left;
      }
      TreeNode currentNode = nodesToProcess.pop();
      traversedNodes.add(currentNode.val);
      root = currentNode.right;
    }
    return traversedNodes;
  }
}
